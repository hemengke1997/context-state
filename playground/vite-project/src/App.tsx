import CounterContainer from './Counter'
import A from './components/A'
import B from './components/B'

function App() {
  return (
    <CounterContainer.Provider>
      <A />
      <B />
      <CounterContainer.Consumer>
        {({ count2, setCount2 }) => (
          <div>
            <span>{count2}</span>
            <button type='button' onClick={() => setCount2((s) => s + 1)}>
              ADD2
            </button>
          </div>
        )}
      </CounterContainer.Consumer>
    </CounterContainer.Provider>
  )
}

export default App
