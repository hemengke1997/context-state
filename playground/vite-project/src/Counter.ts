import { createContainer } from 'context-state'
import React from 'react'

const useCounter = () => {
  const [count, setCount] = React.useState(0)
  const [count2, setCount2] = React.useState(0)

  return {
    count,
    setCount,
    count2,
    setCount2,
  }
}

const CounterContainer = createContainer(useCounter)

export default CounterContainer
