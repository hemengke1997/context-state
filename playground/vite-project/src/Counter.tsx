import React from 'react'
import { createContainer } from 'context-state'

const CounterContainer = createContainer(() => {
  const [count, setCount] = React.useState(0)
  const [count2, setCount2] = React.useState(0)
  return {
    count,
    setCount,
    count2,
    setCount2,
  }
})

export default CounterContainer
