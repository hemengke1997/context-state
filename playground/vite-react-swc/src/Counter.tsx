import { createContainer } from 'context-state'
import React from 'react'

const CounterContainer = createContainer(() => {
  const [count, setCount] = React.useState(0)
  const [count2, setCount2] = React.useState(0)
  return {
    count,
    setCount,
    count2,
    setCount2,
  }
})

export { CounterContainer }
