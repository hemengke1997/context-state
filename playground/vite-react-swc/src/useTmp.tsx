import { useState } from 'react'

export default function useTmp() {
  const [tmp, setTmp] = useState(1)
  return {
    tmp,
    setTmp,
  }
}
