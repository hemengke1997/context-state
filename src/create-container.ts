import { createContext, createElement, type ReactNode, useContext, useRef } from 'react'
import {
  CONTEXT_VALUE,
  type ContextValue,
  EMPTY,
  type EqualityFn,
  type Listener,
  type SelectorFn,
  type UseHookType,
} from './types'
import { pick, shallowEqual, useIsomorphicLayoutEffect, useSafeState } from './utils'

const ErrorText = '[context-state]: Component must be wrapped with <Container.Provider>'

// A hack way for resolving the `Context.Provider` injection
const __ContainerCache__ = new Map<string, ContextValue<any>>()

export function createContainer<Value, InitialValue>(
  useHook: UseHookType<InitialValue, Value>,
  equalityFn: EqualityFn = shallowEqual,
) {
  const _equalityFn = equalityFn

  const Context: React.Context<ContextValue<Value> | symbol> = createContext<ContextValue<Value> | symbol>(EMPTY)

  const Provider = ({ value, children }: { value?: InitialValue; children: ReactNode }) => {
    const inHookValue = useHook(value as InitialValue)
    const valueRef = useRef(inHookValue)
    const contextValue = useRef<ContextValue<Value>>()

    if (!contextValue.current) {
      const listeners = new Set<Listener<Value>>()

      contextValue.current = {
        [CONTEXT_VALUE]: {
          /* "v"alue     */ v: valueRef,
          /* "l"isteners */ l: listeners,
        },
      }
    }

    if (__ContainerCache__.has(useHook.toString())) {
      __ContainerCache__.delete(useHook.toString())
    }

    __ContainerCache__.set(useHook.toString(), contextValue.current)

    useIsomorphicLayoutEffect(() => {
      valueRef.current = inHookValue
      ;(contextValue.current as ContextValue<Value>)?.[CONTEXT_VALUE].l.forEach((listener) => {
        listener(inHookValue)
      })
    }, [inHookValue])

    return createElement(Context.Provider, { value: contextValue.current }, children)
  }

  /**
   * @example
   * import { createContainer } from 'context-state'
   *
   * const Container = createContainer(() => {
   *  const [n, setN] = useState(0)
   * })
   *
   * <Container.Provider>
   *  <Container.Consumer>
   *   {({n}) => <div>{n}</div>}
   *  </Container.Consumer>
   * </Container.Provider>
   */
  const Consumer = ({ children }: { children: (value: Value) => ReactNode }) => {
    return children(useSelector((state) => state))
  }

  /**
   * @example
   * import { useSelector } from 'context-state'
   *
   * export const CounterComponent = () => {
   *   const counter = useSelector(state => state.counter)
   *   return <div>{counter}</div>
   * }
   */
  function useSelector<Selected>(selector: SelectorFn<Value, Selected>, equalityFn?: EqualityFn<Selected>): Selected {
    const eq = equalityFn || _equalityFn

    const context = useContext(Context)

    let contextValue = (context as ContextValue<Value>)?.[CONTEXT_VALUE]

    if (context === EMPTY) {
      const cached = __ContainerCache__.get(useHook.toString())?.[CONTEXT_VALUE]
      if (!cached) {
        throw new Error(ErrorText)
      } else {
        contextValue = cached
      }
    }

    const {
      /* "v"alue     */ v: { current: value },
      /* "l"isteners */ l: listeners,
    } = contextValue

    const [, triggerRender] = useSafeState(0)

    const selected = selector(value)

    const currentRef = useRef<{
      value: Value
      selected: Selected
    }>()

    currentRef.current = {
      value,
      selected,
    }

    useIsomorphicLayoutEffect(() => {
      function checkForUpdates(nextValue: Value) {
        try {
          if (!currentRef.current) {
            return
          }
          if (currentRef.current.value === nextValue) {
            return
          }
          const nextSelected = selector(nextValue)
          if (eq(currentRef.current.selected, nextSelected, shallowEqual)) {
            return
          }
          triggerRender((n) => n + 1)
        } catch {}
      }
      // register listener
      listeners.add(checkForUpdates)
      return () => {
        listeners.delete(checkForUpdates)
      }
    }, [])

    return selected
  }

  /**
   * Syntactic sugar of useSelector
   * @example
   * import { usePicker } from 'context-state'
   *
   * export const CounterComponent = () => {
   *   const { counter } = usePicker(['counter'])
   *   return <div>{counter}</div>
   * }
   */
  function usePicker<Selected extends keyof Value>(
    selected: Selected[],
    equalityFn?: EqualityFn<Pick<Value, Selected>>,
  ): Pick<Value, Selected> {
    return useSelector((state) => pick(state as Required<Value>, selected), equalityFn)
  }

  return {
    Provider,
    Consumer,
    useSelector,
    usePicker,
  }
}
