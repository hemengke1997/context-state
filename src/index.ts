export { shallowEqual, useMemoFn } from './utils'
export { createStore } from './create-store'
