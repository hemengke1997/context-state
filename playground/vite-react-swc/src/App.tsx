import { CounterContainer } from './Counter'
import { A } from './components/A'
import { GlobalContext } from './contexts/GlobalContext'

function App() {
  return (
    <GlobalContext.Provider>
      <CounterContainer.Provider>
        <A />
      </CounterContainer.Provider>
    </GlobalContext.Provider>
  )
}

export default App
