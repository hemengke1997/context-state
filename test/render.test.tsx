import React from 'react'
import { cleanup, fireEvent, render } from '@testing-library/react'
import { afterEach, beforeEach, describe, expect, it, type MockInstance, vitest } from 'vitest'
import { createStore } from '../src'

describe('render spec', () => {
  let consoleErrorSpy: MockInstance

  afterEach(() => {
    cleanup()
    consoleErrorSpy.mockRestore()
  })
  beforeEach(() => {
    consoleErrorSpy = vitest.spyOn(console, 'error').mockImplementation(() => {})
  })

  it('should not render with other', () => {
    const CounterStore = createStore(() => {
      const [count, setCount] = React.useState(0)
      const [total, increment] = React.useReducer((c) => c + 1, 0)
      return {
        count,
        setCount,
        total,
        increment,
      }
    })

    const Counter1 = React.memo(() => {
      const { count, setCount } = CounterStore.useStore(['count', 'setCount'])

      const renderCount = React.useRef(0)

      if (renderCount.current !== count) throw new Error('Inconsistent rendering (1)!')

      renderCount.current += 1

      return (
        <div>
          <span>{count}</span>
          <button type='button' onClick={() => setCount((n) => n + 1)}>
            ADD1
          </button>
        </div>
      )
    })

    const Counter2 = React.memo(() => {
      const { increment } = CounterStore.useStore(['increment'])

      return (
        <div>
          <button type='button' onClick={increment}>
            ADD2
          </button>
        </div>
      )
    })

    const Counter3 = React.memo(() => {
      const { total, increment } = CounterStore.useStore(['total', 'increment'])

      const renderCount = React.useRef(0)

      if (renderCount.current !== total) throw new Error('Inconsistent rendering (3)!')

      renderCount.current += 1
      return (
        <div>
          <span>{total}</span>
          <button type='button' onClick={increment}>
            ADD3
          </button>
        </div>
      )
    })

    const App = () => (
      <CounterStore.Provider>
        <Counter1 />
        <Counter2 />
        <Counter3 />
      </CounterStore.Provider>
    )
    const { getAllByText } = render(<App />)
    expect(() => fireEvent.click(getAllByText('ADD1')[0])).not.toThrow()
    expect(() => fireEvent.click(getAllByText('ADD2')[0])).not.toThrow()
    expect(() => fireEvent.click(getAllByText('ADD3')[0])).not.toThrow()
  })

  it('should throw error in dev', () => {
    const CounterStore = createStore(() => {
      const [count, setCount] = React.useState(0)
      return {
        count,
        setCount,
      }
    })
    const Counter = React.memo(() => {
      const { count } = CounterStore.useStore(['count', 'setCount'])
      return (
        <div>
          <span>{count}</span>
        </div>
      )
    })
    const App = () => <Counter />
    render(<App />)
    expect(consoleErrorSpy).toHaveBeenCalled()
  })

  it('should log error', () => {
    const CounterStore = createStore(() => {
      const [count, setCount] = React.useState(0)
      return {
        count,
        setCount,
      }
    })
    const Counter = React.memo(() => {
      CounterStore.useStore()
      return null
    })
    const App = () => <Counter />
    const { container } = render(<App />)
    expect(container).toMatchSnapshot()
    expect(consoleErrorSpy).toHaveBeenCalled()
  })
})
