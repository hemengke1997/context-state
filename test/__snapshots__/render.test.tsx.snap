// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`render spec > should log error 1`] = `<div />`;
